### ISSUE TITLE
Lexer.lookahead() seems to violate its comment about not modifying the state of the lexer

### ISSUE THREAD

**Issue by haikusw** [graphql/graphql-js#2764]
Apologies in advance if I'm misreading this and for not providing a PR that includes a fix and a test verifying the issue (no Typescript setup or experience).

I'm working on enhancing the Swift [graphql project](https://github.com/GraphQLSwift/GraphQL) which mirrors this js implementation with more complete support for some of the graphql definition language parsing it is currently missing.

In reading the source for the lexer.js [`lookahead` function ](https://github.com/graphql/graphql-js/blob/388cc9a74a514944d9e4b4e913669505aa393045/src/language/lexer.js#L63) I see the comment says:

```
/**
   * Looks ahead and returns the next non-ignored token, but does not change
   * the state of Lexer.
   */
```

However it appears that  `blockstring` support modified the implementation of `readToken` called by `lookahead` such that it can now **_change the state of the lexer_** in violation of the claim in the comment above and the usual expectation for a `lookahead` style function of a lexer.   Not sure if that actually matters in this codebase (_yet_ ;) ), but… 

This modification of the lexer is possible because `lookahead` calls `readToken` passing in the lexer and `readToken` calls `readBlockString` which also passes along the lexer - on [line 192](https://github.com/graphql/graphql-js/blob/388cc9a74a514944d9e4b4e913669505aa393045/src/language/lexer.js#L192).

The passed-in lexer is modified in `readBlockString` on [line 569](https://github.com/graphql/graphql-js/blob/388cc9a74a514944d9e4b4e913669505aa393045/src/language/lexer.js#L569) and again down further on lines 569, 570 and 578.

Thought I'd mention it so that someone working on the project can write a quick test and decide if this is the bug lurking in the shadows it appears to potentially be.  

Cheers.

P.S.  In my implementation of `lookahead` I simply save off `line` and `lineStart` at the beginning and then restore before returning  from `lookahead`.  Somewhat fragile in that any future changes to `readToken` or `readBlockString` that modify additional fields of the lexer may not know to update the save/restore code in `lookahead`, so a collection of unit tests to catch this if it happens is something I plan to add.  Considered duplicating the lexer but that has performance considerations.  Anyway, seems likely something similar will work in this js version.

**Comment by IvanGoncharov** [comment]
@haikusw Yes, you absolutely right. Support of block strings is a relatively new addition compared to the rest of GraphQL so it was implemented somewhat ad-hoc.
We definitely need to return to previous behavior and since we need to rewrite Lexer to support comments as part of the AST we can address this issue at the same time.

**Comment by haikusw** [comment]
Thanks for the confirmation.  Not sure you schedule for those bigger changes, but something like the save/restore strategy I did for the Swift version of `lookahead()` here:

https://github.com/GraphQLSwift/GraphQL/blob/023af5acd113dba5502d2357558bc91cc71c8d71/Sources/GraphQL/Language/Lexer.swift#L99

(Lines 101-102 and then 116-117)

might be a decent short-term fix if those bigger things are out on the horizon some.

peace.

**Comment by yaacovCR** [comment]
Closed by #4294
